### PR TITLE
Do without a catenable queue

### PR DESCRIPTION
**Pull request by treeowl**
We can use `toView` to turn a `SeqT` into something we can
shove into a plain queue. Good idea? Bad idea? We'll need
benchmarks to tell.

### CODE REVIEW

**Comment by treeowl** [comment]
There's a bug in both the type-aligned and non-type-aligned `FastQueue` implementations, causing invariant failures in their internals. I copied the non-type-aligned version into this repo, made a tweak that gets the tests passing, and added a fake existential that should clarify that the schedule is only a schedule. We can and should experiment with other queues.

**Comment by dagit** [comment]
> 
> 
> We can use `toView` to turn a `SeqT` into something we can
> shove into a plain queue. Good idea? Bad idea? We'll need
> benchmarks to tell.

We do have benchmarks! Are you able to run them?

**Comment by dagit** [comment]
I guess this means we should make a version, even if only temporarily for benchmarking, that is parametric in the underlying queue implementation. I'm imaginging that `toView/fromView` and maybe `single/singleton` (not sure what else) could be in a type class and we'd just write instances for the things we're interested in benchmarking.

**Comment by dagit** [comment]
> There's a bug in both the type-aligned and non-type-aligned `FastQueue` implementations, causing invariant failures in their internals.

Ugh. I'm glad you caught that!

**Comment by treeowl** [comment]
The bug doesn't seem to affect the catenable queue, because it only happens with empty queues, and the catenable queues represent those differently. I don't *think* we want to try to be parametric in the queue type, unless we do it with something like `backpack` (beyond my current pay grade, and incompatible with `stack`), because that'll make inlining and specialization even harder to figure out.

I haven't tried running the benchmarks yet, and I'm not sure what I'd need to do to make my old laptop quiet enough to get consistent results. Would you mind giving it a whirl or two?

**Comment by treeowl** [comment]
I have no fscking clue why Haddock explodes on ghc < 7.8.

**Comment by treeowl** [comment]
It was this Haddock bug: https://github.com/haskell/haddock/issues/242.

It looks like maybe [the workaround in `lens`](https://github.com/ekmett/lens/commit/89b27760c6d6b083e14d0c07f2f832596729c141) is to make sure    `haddock:  <7.6 || >=7.8`, but I dunno how to do that in our CI.

**Comment by treeowl** [comment]
OK, I successfully nixed Haddock for the old GHC versions that were buggy. Windows is still goofing things up :-(.

**Comment by treeowl** [comment]
Oh, the Windows thing might actually be my fault this time, thanks to differences between shells.

**Comment by treeowl** [comment]
@dagit, I don't know how to Windows. Give me a clue about how to fix that syntax error?

File: src/Control/Monad/Logic/Sequence/Internal.hs [hunk 1]
```diff
@@ -181,7 +181,7 @@ toView (SeqT s) = case S.viewl s of
   S.EmptyL -> return Empty
   h S.:< t -> h >>= \x -> case x of
     Empty -> toView (SeqT t)
-    hi :< SeqT ti -> return (hi :< SeqT (ti S.>< t))
+    hi :< SeqT ti -> return (hi :< SeqT (ti S.|> toView (SeqT t)))
```
Comment (treeowl) [inline src/Control/Monad/Logic/Sequence/Internal.hs:184]: This makes me nervous. Might well be the wrong asymptotics.

Comment (treeowl) [reply src/Control/Monad/Logic/Sequence/Internal.hs:184]: Actually... Probably fine. Will need to analyze further.

Comment (treeowl) [reply src/Control/Monad/Logic/Sequence/Internal.hs:184]: No, forget it...

**Comment by dagit** [comment]
> 
> 
> @dagit, I don't know how to Windows. Give me a clue about how to fix that syntax error?

Sorry. I don't have a clue either.

File: src/Control/Monad/Logic/Sequence/Internal.hs (hunk 1 above)
Comment (dagit) [reply src/Control/Monad/Logic/Sequence/Internal.hs:184]: I'm late to the party. What did you learn? Bad asymptotics?

Comment (treeowl) [reply src/Control/Monad/Logic/Sequence/Internal.hs:184]: Yeah, I'm pretty sure it's wrong. I do want to go back to the idea of unpacking the simple queue into the catenable one. And maybe try adding some special constructors for certain things; not sure.

**Pull request closed without merging**